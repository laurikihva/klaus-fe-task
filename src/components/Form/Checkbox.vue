<template>
  <label :class="labelClasses">
    <span class="relative block w-[16px] h-[16px]">
      <input :id v-model="model" class="sr-only peer" :name :checked="model" type="checkbox" :disabled />
      <span :class="checkboxClasses" />
      <span v-if="!indeterminate" :class="checkmarkClasses">
        <Icon name="check" size="small" />
      </span>
      <span v-else :class="indeterminateClasses" />
    </span>
    <slot />
  </label>
</template>

<script lang="ts" setup>
import { toRefs, computed } from "vue";

import Icon from "../Icon.vue";

export interface ICheckboxProps {
  id: string;
  name?: string;
  indeterminate?: boolean;
  disabled?: boolean;
}

const props = defineProps<ICheckboxProps>();
const { indeterminate, disabled } = toRefs(props);

const model = defineModel<boolean>({ required: true });

const labelClasses = computed(() => {
  const base = "flex items-center gap-2";
  const isDisabled = disabled.value ? "cursor-not-allowed opacity-50" : "cursor-pointer";

  return [base, isDisabled];
});
const checkboxClasses = computed(() => {
  const base = `
    block
    w-full
    h-full
    rounded-[3px]
    border
    border-gray-200
    peer-focus:shadow-primary
  `;
  const checkedClasses = indeterminate.value
    ? "bg-blue-400 border-blue-400"
    : "peer-checked:bg-blue-400 peer-checked:border-blue-400";

  return [base, checkedClasses];
});

const sharedIconClasses = "absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2";

const checkmarkClasses = computed(() => {
  const base = "text-white";
  return [base, sharedIconClasses];
});
const indeterminateClasses = computed(() => {
  const base = "bg-white rounded-full w-[9px] h-[2px]";
  return [base, sharedIconClasses];
});
</script>
