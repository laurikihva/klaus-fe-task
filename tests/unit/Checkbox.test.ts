import { mount } from "@vue/test-utils";

import Checkbox from "@/components/Form/Checkbox.vue";

const wrapper = mount(Checkbox, {
  props: {
    id: "my-check",
    modelValue: false,
    disabled: true,
  },
  slots: {
    default: "Checkbox",
  },
});

describe("Loads and displays Checkbox component", () => {
  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("gets checked on modelValue change", async () => {
    expect(wrapper.props("modelValue")).toBe(false);
    await wrapper.setProps({ disabled: false, modelValue: true });
    expect(wrapper.props("modelValue")).toBe(true);
  });

  it("contains indeterminate icon", async () => {
    await wrapper.setProps({ indeterminate: true });

    expect(wrapper.find("svg").exists()).toBeFalsy();
    expect(wrapper.find({ ref: "indeterminateRef" }).classes("bg-white")).toBe(true);
    expect(wrapper.find({ ref: "indeterminateRef" }).classes("rounded-full")).toBe(true);
  });
});
