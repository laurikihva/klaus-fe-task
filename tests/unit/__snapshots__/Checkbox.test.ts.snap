// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loads and displays Checkbox component > renders correctly 1`] = `
"<label class="flex items-center gap-2 cursor-not-allowed opacity-50"><span class="relative block w-[16px] h-[16px]"><input id="my-check" class="sr-only peer" type="checkbox" disabled=""><span class="block
    w-full
    h-full
    rounded-[3px]
    border
    border-gray-200
    peer-focus:shadow-primary peer-checked:bg-blue-400 peer-checked:border-blue-400"></span><span class="text-white absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2"><svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 10 8" class="w-[10px] h-[10px]"><path d="M7.853.808a1 1 0 0 1 1.488.04l.066.078a1 1 0 0 1-.047 1.348l-4.595 4.69a1 1 0 0 1-1.429 0L.654 4.224A1 1 0 0 1 .62 2.861l.047-.053a1 1 0 0 1 1.477-.019L4.05 4.824z"></path></svg></span></span>Checkbox</label>"
`;
